File: prioritarr/webhooks/plex_event.py
from __future__ import annotations

from dataclasses import dataclass

from prioritarr.database import Database


@dataclass
class WatchedEvent:
    show_title: str
    plex_show_key: str
    season: int
    episode: int
    rating_key: str


def parse_tautulli_watched(data: dict) -> WatchedEvent:
    """Extract fields from a Tautulli webhook payload for a watched episode."""
    return WatchedEvent(
        show_title=data.get("grandparent_title", ""),
        plex_show_key=data.get("grandparent_rating_key", ""),
        season=int(data.get("parent_media_index", 0) or 0),
        episode=int(data.get("media_index", 0) or 0),
        rating_key=data.get("rating_key", ""),
    )


def handle_watched(sonarr_series_id: int, db: Database) -> None:
    """Invalidate the priority cache for the given series and log the action."""
    db.invalidate_priority_cache(sonarr_series_id)
    db.append_audit(
        action="cache_invalidated",
        series_id=sonarr_series_id,
        details={"reason": "tautulli_watched"},
    )

File: prioritarr/backend/src/main/kotlin/org/yoshiz/app/prioritarr/backend/webhooks/PlexEvent.kt
package org.yoshiz.app.prioritarr.backend.webhooks

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import org.yoshiz.app.prioritarr.backend.database.Database

data class WatchedEvent(
    val showTitle: String,
    val plexShowKey: String,
    val season: Int,
    val episode: Int,
    val ratingKey: String,
)

fun parseTautulliWatched(data: JsonObject): WatchedEvent {
    fun intField(key: String): Int =
        data[key]?.jsonPrimitive?.let {
            it.intOrNull ?: it.contentOrNull?.toIntOrNull()
        } ?: 0

    return WatchedEvent(
        showTitle = data["grandparent_title"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        plexShowKey = data["grandparent_rating_key"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        season = intField("parent_media_index"),
        episode = intField("media_index"),
        ratingKey = data["rating_key"]?.jsonPrimitive?.contentOrNull.orEmpty(),
    )
}

/** Invalidate priority cache + log audit. Mirrors handle_watched in python. */
fun handleWatched(seriesId: Long, db: Database) {
    db.invalidatePriorityCache(seriesId)
    db.appendAudit(
        action = "cache_invalidated",
        seriesId = seriesId,
        details = buildJsonObject { put("reason", "tautulli_watched") },
    )
}
